File: dynamic_programming_python/armstrong_number.py
'''
In this code, the function is_armstrong_number() takes an integer n as input and 
checks whether it is an Armstrong number. It does this by converting the number 
to a string and iterating over its digits. For each digit, it raises it to the 
power of the total number of digits and adds it to the armstrong_sum. Finally, it 
compares the armstrong_sum with the original number n to determine if it is an 
Armstrong number.

You can test the code by entering a number, and it will output whether it is an 
Armstrong number or not.
'''
def is_armstrong_number(n):
    # Convert the number to a string to iterate over its digits
    digits = str(n)
    num_digits = len(digits)
    armstrong_sum = 0

    for digit in digits:
        armstrong_sum += int(digit) ** num_digits

    return armstrong_sum == n

# Example usage
number = int(input("Enter a number: "))
if is_armstrong_number(number):
    print(number, "is an Armstrong number")
else:
    print(number, "is not an Armstrong number")

File: dynamic_programming_cpp/armstrong_number.cpp
#include <iostream>
#include <cmath>

/*
In this code, the isArmstrongNumber() function takes an integer n as input and 
checks whether it is an Armstrong number. It does this by counting the number 
of digits in the original number and then calculating the Armstrong sum. The 
Armstrong sum is obtained by raising each digit to the power of the number of 
digits and adding them together. Finally, the function compares the Armstrong 
sum with the original number to determine if it is an Armstrong number.

In the main() function, the user is prompted to enter a number. The 
isArmstrongNumber() function is called with the entered number, and 
the result is printed accordingly.
*/

using namespace std;

bool isArmstrongNumber(int n) {
    int originalNumber = n;
    int numDigits = 0;
    int armstrongSum = 0;

    // Count the number of digits
    while (originalNumber != 0) {
        originalNumber /= 10;
        numDigits++;
    }

    // Calculate the Armstrong sum
    originalNumber = n;
    while (originalNumber != 0) {
        int digit = originalNumber % 10;
        armstrongSum += pow(digit, numDigits);
        originalNumber /= 10;
    }

    return (armstrongSum == n);
}

int main() {
    int number;
    cout << "Enter a number: ";
    cin >> number;

    if (isArmstrongNumber(number))
        cout << number << " is an Armstrong number." << endl;
    else
        cout << number << " is not an Armstrong number." << endl;

    return 0;
}

